railties/lib/rails/secrets.rb

```diff
@@ -0,0 +1,111 @@
+require "yaml"
+
+module Rails
+  # Greatly inspired by Ara T. Howard's magnificent sekrets gem. 😘
+  class Secrets # :nodoc:
+    class MissingKeyError < RuntimeError
+      def initialize
+        super(<<-end_of_message.squish)
+          Missing encryption key to decrypt secrets with.
+          Ask your team for your master key and put it in ENV["RAILS_MASTER_KEY"]
+        end_of_message
+      end
+    end
+
+    @read_encrypted_secrets = false
+    @root = File # Wonky, but ensures `join` uses the current directory.
+
+    class << self
+      attr_writer   :root
+      attr_accessor :read_encrypted_secrets
+
+      def parse(paths, env:)
+        paths.each_with_object(Hash.new) do |path, all_secrets|
+          require "erb"
+
+          secrets = YAML.load(ERB.new(preprocess(path)).result) || {}
+          all_secrets.merge!(secrets["shared"].deep_symbolize_keys) if secrets["shared"]
+          all_secrets.merge!(secrets[env].deep_symbolize_keys) if secrets[env]
+        end
+      end
+
+      def generate_key
+        cipher = new_cipher
+        SecureRandom.hex(cipher.key_len)[0, cipher.key_len]
```

Comment:
This is limiting the key's entropy. The output of

```ruby
SecureRandom.hex(n).size # => 2n
```

is twice as long because `n` is telling `SecureRandom` "give me an output that is equivalent to `n` random bytes". Hex encoding encodes every byte as two characters, so the resulting hex string would be twice as long.

For the default AES-256 `n = 32`, and by cutting the 64 hex bytes down to 32 again, the entropy in that value is now only equivalent to 16 random bytes anymore. This makes brute forcing the key much easier: Instead of 2^256 now there are only 2^128 possible keys.

I would suggest simply using the full output of `SecureRandom.hex(cipher.key_len)` and decoding the resulting 64 hex bytes when the key is read from the file, yielding the 32-byte binary key.